W-10742407: Documentation section about How to document your DW function (#115)
mulesoft/docs-dataweave: modules/ROOT/pages/dataweave-create-module.adoc

* ----
* YOUR CODE
* ----
*
* ==== Input (optional section)
*
* The input description (optional)
*
* [source,%The language%,linenums] (optional)
* ----
* YOUR CODE
* ----
*
* ==== Output (optional section)
*
* %The output description% (optional)
*
* [source,%The language%,linenums] (optional)
* ----
* YOUR CODE
* ----
*/
----

==== Description Section

This section provides a description of the function. The section starts at the top of the comments and extends to the line before the first section heading. If there is no section header, the description extends to the end of the comments for the function.

A description consists of the following parts:

* Short description: The first paragraph, which provides the primary description of the function.
* Long description: Any additional information about the function. This optional description begins two line breaks below the short description.

See the <<example-documentation, example>> that includes a description with sections for the parameters and examples.

Each part is useful for auto-generated documentation.

==== Parameters Section

The optional `=== Parameters` section describes parameters of the function.

> The template uses Asciidoc table format to document each parameter in a table row.

Reviewer: ```suggestion
The template uses AsciiDoc table format to document each parameter in a table row.
```

....
 [%header, cols="1,13"]
 |===
 | Name | Description
 | `yourParameter1` | Your description here.
 | `yourParameter2` | Your description here.

>  |===

Author: For this section we've 2 templates:
- Legacy format, without parameter type:

[%header, cols="1,13"]
|===
| Name | Description
| `yourParameter1` | Your description here.
| `yourParameter2` | Your description here.
|===

- New format, with parameter type:
[%header, cols="1,1,3"]
|===
| Name | Type | Description
| `yourParameter1` | The parameter type | Your description here.
| `yourParameter2` | The parameter type | Your description here.
|===

IMO we should document the 2nd template or both. Because if you use the DataWeave VSCode Extension to generate your docs, you will end with the 2nd template.

....

==== Example Section

You can use zero or more `=== Example` sections to provide any examples needed to illustrate how your function works.

This section contains the following optional subsections:

* `==== Input` for input to the DataWeave script.
* `==== Source` for the DataWeave script.

> * `==== Input` for output generated by the script.

Author: Replace `==== Input` by `==== Output`

All subsections follow the same template, which includes an optional description and an optional code section:

....
Your section description

[source,%The language%,linenums]
----
YOUR CODE
----
....

[[example-documentation]]

> The following example shows the xref:dw-core-functions-log.adoc[log] function documentation.

Reviewer: ```suggestion
The following example shows the https://docs.mulesoft.com/dataweave/2.4/dw-core-functions-log[log] function documentation.
```

....
/**
* Without changing the value of the input, `log` returns the input as a system
* log. So this makes it very simple to debug your code, because any expression or subexpression can be wrapped
* with *log* and the result will be printed out without modifying the result of the expression.
* The output is going to be printed in application/dw format.
* (1)
*
* The prefix parameter is optional and allows to easily find the log output.
*
*
* Use this function to help with debugging DataWeave scripts. A Mule app
* outputs the results through the `DefaultLoggingService`, which you can see
* in the Studio console.
*
* === Parameters
*
* [%header, cols="1,13"]
* |===
* | Name | Type | Description
* | `prefix` | `String` | An optional string that typically describes the log.

> * | `value` | `TODO` | The value to log.

Reviewer: @mlischetti what type should we put here?

Author: In this case, T is the parameter. Because it's a parameterized type

> * | `value` | `TODO` | The value to log.

Reviewer: ```suggestion
* | `value` | `T` | The value to log.
```

*
* [source,DataWeave,linenums]
* ----
* %dw 2.0
* output application/json
* ---
* log("WARNING", "Houston, we have a problem")
* ----
*
* ==== Output
*
* `Console Output`
*
* [source,XML,linenums]
* ----
* "WARNING - Houston, we have a problem"
* ----
*
* `Expression Output`
*
* [source,XML,linenums]
* ----
* "Houston, we have a problem"
* ----
*
* === Example
*
* This example shows how to log the result of expression `myUser.user` without modifying the
* original expression `myUser.user.friend.name`.
*
* ==== Source
*
* [source,DataWeave,linenums]
* ----
* %dw 2.0
* output application/json
*
* var myUser = {user: {friend: {name: "Shoki"}, id: 1, name: "Tomo"}, accountId: "leansh" }
* ---
* log("User", myUser.user).friend.name
* ----
*
* ==== Output
*
* `Console output`
*
* [source,console,linenums]
* ----
* User - {
*   friend: {
*     name: "Shoki"
*   },
*   id: 1,
*   name: "Tomo"
* }
* ----
*
* `Expression Output`
*
* [source,DataWeave,linenums]
* ----
* "Shoki"
* ----
*/
....

=== Documenting Your DataWeave Modules

The DataWeave module description does not follow a structured template.

The only requirement is to place the documentation above the DataWeave version tag.

> The following example documents the xref:dw-core.adoc[dw::Core] module:

Reviewer: ```suggestion
The following example documents the https://docs.mulesoft.com/dataweave/2.4/dw-core[dw::Core] module:
```

[source,dataweave,linenums]
----
/**
 * This module contains core DataWeave functions for data transformations.
 * It is automatically imported into any DataWeave script. For documentation
 * on DataWeave _1.0_ functions, see
 * https://docs.mulesoft.com/dataweave/1.2/dataweave-operators[DataWeave Operators].
 */
%dw 2.0
----

=== Documenting Your Annotation and Types

Annotation and Types documentation do not follow a structured template.

The only requirement is to write the descriptions above the Annotation or Type definition.

The following example documents the `@StreamCapable` annotation.

[source,dataweave,linenums]
----
/**
* Annotation that marks a parameter as stream capable, which means that this
* field will consume an array of objects in a forward-only manner.
*/
@AnnotationTarget(targets = ["Parameter", "Variable"])
annotation StreamCapable()
----

> The following example documents the `EncodingSettings` type found in the xref:dw-dataformat.adoc[dw::extension::DataFormat] module.

Reviewer: ```suggestion
The following example documents the `EncodingSettings` type found in the https://docs.mulesoft.com/dataweave/2.4/dw-dataformat[dw::extension::DataFormat] module.
```
